src/NServiceBus.Persistence.AzureStorage/SagaPersisters/AzureSagaPersister.cs

```diff
@@ -101,6 +99,29 @@ public Task Update(IContainSagaData sagaData, SynchronizedStorageSession session
                     }
                 }
             }
+
+            await RemoveSecondaryIndex(sagaData, table).ConfigureAwait(false);
+        }
+
+        static async Task RemoveSecondaryIndex(IContainSagaData sagaData, CloudTable table)
+        {
+            PartitionRowKeyTuple secondaryIndexKey;
+            if (secondaryIndexKeys.TryGetValue(sagaData, out secondaryIndexKey))
+            {
+                var e = new TableEntity
+                {
+                    ETag = "*"
+                };
+                secondaryIndexKey.Apply(e);
+                try
+                {
+                    await table.ExecuteAsync(TableOperation.Delete(e)).ConfigureAwait(false);
+                }
+                catch
+                {
+                    // this is only secondary index removal. If it fails, an oprhaned entity is the only cost
```

Comment:
What's the probability of having this entity orphaned? IMO it's highly unlikely that it will happen as we do it after removing the primary. The problem is that this is the very last operation for this saga. It won't be called no more. We'd need a separate tool for GC. We could mark the secondary that the primary is about to be deleted, delete the primary and try to delete the secondary (as in the above). When it fails, this GC will scan for the seconaries being marked for removal and having their primaries removed. If that's the case, remove it.
That's the only way I see it. If you see another, please do share it.